### ISSUE TITLE
Implementing play/pause and other improvements

### ISSUE THREAD

**Issue by marioortizmanero** [SwagLyrics/SwSpotify#3]
(Continuation of https://github.com/marioortizmanero/spotify-music-videos/issues/10)

The main loop is just an endless loop that calls a [function](https://github.com/marioortizmanero/spotify-music-videos/blob/next/spotify_videos/api/linux.py#L122) you [provide](https://github.com/marioortizmanero/spotify-music-videos/blob/next/spotify_videos/api/linux.py#L50) with data of events happening when the dbus bus detects a change like pause or a new song (you can check my dbus module to see how it works). I use it in Spotify Music Videos to control the player and know when songs end.

I will be able to do a couple PRs soon-ish but I'm not really sure how you want it done. Like what I have inside my implementation?

**Comment by aadibajpai** [comment]
Um, I think it would be cool if the functions currently exposed remain the same but the underlying logic is improved for the play/pause thing as well as not opening multiple dbus sessions. I was thinking a new Exception like SpotifyPaused could be defined that inherits from SpotifyNotRunning maybe to handle that. I'm not sure how play/pause can be detected on macOS btw.

**Comment by marioortizmanero** [comment]
i'm not sure if SpotifyPaused should be an exception... How about 3 different classes with `@property` that call whatever is needed to obtain info? That way the dbus session can be stored inside the object, for example, and is more solid. This is done in almost every other API.

**Comment by marioortizmanero** [comment]
I've finally managed to install a macOS VM. Here's what returning the playback status would be like:

```python
def get_info_mac():
    """
    isPlaying may be 'playing' or 'paused'
    """

    from Foundation import NSAppleScript
    apple_script_code = """
    getCurrentlyPlayingTrack()
    on getCurrentlyPlayingTrack()
        tell application "Spotify"
                set isPlaying to the player state as string
                set currentArtist to artist of current track as string
                set currentTrack to name of current track as string
                return {currentArtist, currentTrack, isPlaying}
        end tell
    end getCurrentlyPlayingTrack
    """

    s = NSAppleScript.alloc().initWithSource_(apple_script_code)
    x = s.executeAndReturnError_(None)
    a = str(x[0]).split('"')
    is_playing = True if a[5].lower() == 'playing' else False
    return a[3], a[1], is_playing

print(get_info_mac())
```

**Comment by aadibajpai** [comment]
Sorry I haven't been more active lately, it's the weekend finally so I think I'll be able to get to this now. 

The reason I was preferring an exception over an attribute was that if song isn't playing then the other two variables being returned are meaningless and quite opposite in function. On the other hand if you do an inherited exception, then the user could do something like:

```python
except SpotifyNotRunning as e:
    print(e)
    ...
```
Now if the exception is spotify paused then that would be reflected in the `e` and so would if spotify isn't even open. If someone wants to do different things based on whether it's paused or the app is closed, they would have the individual exceptions to handle.

I'll take a closer look at this tonight, this is just the design I had in mind for this. Would love to hear what you think. In any case, the core logic remains the same so PR is enormously useful 😄

**Comment by marioortizmanero** [comment]
Rather than playing with exceptions if what you want is to avoid returning useless data, maybe the title and artist could be set to a global variable so that it still returns the same metadata for both of them, alongside with the `is_playing` variable. I think it's easier to handle `is_playing` being true or false than catching an exception. Just take a look to the PR and tell me what you think :). I've implemented a lot of stuff.

**Comment by aadibajpai** [comment]
A separate thing but I was wondering if you were interested in adding support for the Spotify Web Player as well? It can be done and we have a Chrome extension that probably needs to be updated and integrated here.

**Comment by marioortizmanero** [comment]
Sure but that sounds a bit more complicated, I'd do it on the next release after this. Also because the web API I'm interested on is still under development.

Until this API is finished I'd make sure this version is ready for release.

**Comment by aadibajpai** [comment]
> Sure but that sounds a bit more complicated, I'd do it on the next release after this. Also because the web API I'm interested on is still under development. Here it is https://github.com/felix-hilden/spotipy
> 
> Until this API is finished I'd make sure this version is ready for release.

Yeah definitely not for this release, but for SwSpotify 2.0 maybe

**Comment by aadibajpai** [comment]
Hmm looks like that requires a token, the extension we have doesn't need one. It's pretty similar to how SwSpotify works in all OSes.

**Comment by marioortizmanero** [comment]
Yup that's the downside. It can be a nice idea to have it as an opt-in, which is what I do on my module. But it might also not be worth it.

**Comment by aadibajpai** [comment]
> Yup that's the downside. It can be a nice idea to have it as an opt-in, which is what I do on my module. But it might also not be worth it.

I was talking about spotipy, usually getting token involves using the API which is considerably slower than our methods here. The extension (https://github.com/SwagLyrics/SwagLyrics-Chrome-Extension) doesn't need it and it sends info via a cors request

**Comment by marioortizmanero** [comment]
Oh If you're talking about not using the official API it might take me a while because I'm not too experienced with web scrapping. But I'm not sure how reliable it would be.

**Comment by aadibajpai** [comment]
> Oh If you're talking about not using the official API it might take me a while because I'm not too experienced with web scrapping. But I'm not sure how reliable it would be.

That extension does work but it basically needs to be overhauled and integrated, some work was started but it never finished https://github.com/SwagLyrics/SwagLyrics-For-Spotify/pull/962

I was hoping to do it but I would need to learn a bit about extensions and js so I've been delaying it. That would probably make this library pretty supreme

**Comment by marioortizmanero** [comment]
Definitely sounds like a great idea. I'd rather do it after your overhaul is finished to have something more solid to look at, but if I have time I can try doing it before that.

**Comment by aadibajpai** [comment]
> Definitely sounds like a great idea. I'd rather do it after your overhaul is finished to have something more solid to look at, but if I have time I can try doing it before that.

The dude who started it is out of the picture lol so it's pretty much in need of revival. It would be nice to have someone look at it, yes.

**Comment by aadibajpai** [comment]
This done? :D

**Comment by marioortizmanero** [comment]
Yep! Anything else that needs to be done before the release? I'll do some testing on macOS today.

**Comment by aadibajpai** [comment]
> Yep! Anything else that needs to be done before the release? I'll do some testing on macOS today.

Yup, basically looking to edit that blanket exception for macOS, get it tested once on all platforms and then make a new release!

> Edit: sorry but I won't be able to test it until the end of the weekend

No problem! Take your time :)

**Comment by marioortizmanero** [comment]
Bad news: my VM was a 60-day free trial of VMware Workstation. This took too long and now it's over :(

Can you download it yourself? I'm not sure of how I could keep using it.

**Comment by aadibajpai** [comment]
> Bad news: my VM was a 60-day free trial of VMware Workstation. This took too long and now it's over :(
> 
> Can you download it yourself? I'm not sure of how I could keep using it.
> 
> It does work on my Windows setup though,

No problem, I had asked a friend who had a mac to test and he said IndexError, I'll just get it checked once again and we should be good to go. Can you confirm it works on linux?

**Comment by marioortizmanero** [comment]
Yup, works perfectly on Linux. Also, after the new version is released, I'll upload it to the [AUR](https://aur.archlinux.org/) and let you know. It's just a repository to download the package from Arch Linux.

**Comment by marioortizmanero** [comment]
Yup, works perfectly on Linux. What's the new version number going to be?

**Comment by aadibajpai** [comment]
> Yup, works perfectly on Linux. What's the new version number going to be?

I'm trying to see if it's backward compatible, if it is then v1.1.0 otherwise v2.0.0

**Comment by aadibajpai** [comment]
```
>>> from SwSpotify import spotify
>>> spotify.get_info_mac()
Traceback (most recent call last):
  File "/Users/matejplavevski/Downloads/SwSpotify/SwSpotify/spotify.py", line 107, in get_info_mac
    if a[5].lower != 'playing':
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/matejplavevski/Downloads/SwSpotify/SwSpotify/spotify.py", line 110, in get_info_mac
    raise SpotifyClosed
SwSpotify.SpotifyClosed: Spotify appears to be closed at the moment.
```
Output on a Mac with spotify closed

**Comment by aadibajpai** [comment]
Done, thanks again for all your help and sorry for the delay!

**Comment by marioortizmanero** [comment]
Great!! If you want to discuss further the Spotify website scrapper you can open up a new issue. But I won't be able to get my hands in it soon.
